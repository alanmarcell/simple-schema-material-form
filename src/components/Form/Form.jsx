import React from 'react';
import PropTypes from 'prop-types';
import { pathOr, merge, isNil, isEmpty, filter } from 'ramda';
import { compose, withStateHandlers, pure, withState } from 'recompact';
import { FormControl } from 'material-ui/Form';
import { Button } from 'material-ui';
import handleSubmit from './submitHandler';

const hasError = (errors = [], fieldName) => {
  if (isNil(errors)) {
    return false;
  }

  const getError = error => {
    if (error.name === fieldName) {
      return error;
    }

    return null;
  };

  const fieldError = filter(getError, errors);

  if (isEmpty(fieldError)) {
    return false;
  }

  return true;
};

const getChildProps = ({ doc, setDoc, errors }) => Child => {
  const { fieldName } = Child.props;
  const value = pathOr('', [fieldName], doc);

  const childProps = {
    ...Child.props,
    setDoc,
    doc,
    error: hasError(errors, fieldName),
    value,
    label: fieldName,
  };

  return React.createElement(
    Child.type,
    childProps,
    null,
  );
};

/** Our Simple Form Test Utility */
const SimpleForm = (props) => {
  const { doc, setDoc, errors } = props;
  const setChildProps = getChildProps({ doc, setDoc, errors });
  return (
    <FormControl >
      {React.Children.map(props.children, setChildProps)}
      <Button onClick={() => handleSubmit(props)} >Submit</Button>
    </FormControl>
  );
};

/** This is our base enhancer to handle doc changes */
const enhance = compose(
  pure,
  withState('errors', 'setErrors', []),
  withStateHandlers(
    /** this our initial doc */
    { doc: {} },
    {
      /** this is the doc state handler */
      setDoc: ({ doc }) => value => {
        const merged = merge(doc, value);
        return {
          doc: merged,
        };
      },
    },
  ),
);

const EnhancedSimpleForm = enhance(SimpleForm);

SimpleForm.defaultProps = {
  /** This is our fields */
  // children: null,
};

SimpleForm.propTypes = {
  /** doc defines the form state */
  doc: PropTypes.object.isRequired,

  /** handle the form state */
  setDoc: PropTypes.func.isRequired,

  children: PropTypes.node.isRequired,
};

export { SimpleForm };
export default EnhancedSimpleForm;
