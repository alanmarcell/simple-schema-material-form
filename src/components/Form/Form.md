# Form example

```jsx
<Form name="Hello" >
  <TextInput fieldName="name" />
</Form>
```
