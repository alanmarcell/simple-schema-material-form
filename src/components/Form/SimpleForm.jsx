import React from 'react';
import PropTypes from 'prop-types';
import { omit, pathOr, merge } from 'ramda';
import { compose, withStateHandlers, pure } from 'recompact';
import { Button, TextField } from 'material-ui';

const getChildProps = ({ doc, setDoc }) => Child => {
  const { fieldName } = Child.props;
  const value = pathOr('', [fieldName], doc);

  const childProps = {
    ...Child.props,
    setDoc,
    doc,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };

  return React.createElement(
    Child.type,
    childProps,
    null,
  );
};

export const TextInput = props => {
  const { doc, setDoc, fieldName } = props;
  const value = pathOr('', [fieldName], doc);
  const cProps = omit(['fieldName', 'setDoc', 'doc'], props);

  const inputProps = {
    ...cProps,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };
  return <TextField {...inputProps} />;
};

/** Our Simple Form Test Utility */
const SimpleForm = (props) => {
  const { doc, setDoc, onSubmit } = props;
  const setChildProps = getChildProps({ doc, setDoc });
  return (
    <form >
      {React.Children.map(props.children, setChildProps)}
      <Button onClick={() => onSubmit(doc)} >Submit</Button>
    </form>
  );
};

/** This is our base enhancer to handle doc changes */
const enhance = compose(
  pure,
  withStateHandlers(
    /** this our initial doc */
    { doc: {} },
    {
      /** this is the doc state handler */
      setDoc: ({ doc }) => value => {
        const merged = merge(doc, value);
        return {
          doc: merged,
        };
      },
    },
  ),
);

const EnhancedSimpleForm = enhance(SimpleForm);

SimpleForm.defaultProps = {
  /** This is our fields */
  // children: null,
};

SimpleForm.propTypes = {
  /** doc defines the form state */
  doc: PropTypes.object.isRequired,

  /** handle the form state */
  setDoc: PropTypes.func.isRequired,

  children: PropTypes.array.isRequired,
};

export default EnhancedSimpleForm;
