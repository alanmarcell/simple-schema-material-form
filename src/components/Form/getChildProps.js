import React from 'react';
import { pathOr, isNil, isEmpty, filter } from 'ramda';

const getFieldError = fieldName => error => {
  if (error.name === fieldName) {
    return error;
  }
  return null;
};

const hasError = (errors = [], fieldName) => {
  if (isNil(errors)) {
    return false;
  }

  return !isEmpty(filter(getFieldError(fieldName), errors));
};


const getChildProps = ({ doc, setDoc, errors }) => Child => {
  const { fieldName } = Child.props;
  const value = pathOr('', [fieldName], doc);

  const childProps = {
    ...Child.props,
    setDoc,
    doc,
    error: hasError(errors, fieldName),
    value,
    label: fieldName,
  };

  return React.createElement(
    Child.type,
    childProps,
    null,
  );
};

export default getChildProps;
