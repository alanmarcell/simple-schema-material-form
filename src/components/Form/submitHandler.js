import validateDoc from './validateForm';

const handleSubmit = ({
  buildDocToSave = doc => doc,
  buildDocToValidate = doc => doc,
  ...props
}) => {
  const docIsValid = validateDoc(buildDocToValidate(props));
  const {
    onSubmit,
  } = props;

  if (docIsValid) {
    const docToSave = buildDocToSave(props.doc);

    onSubmit(docToSave);
  }
};

export default handleSubmit;
