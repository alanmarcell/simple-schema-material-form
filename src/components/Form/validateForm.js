const validateDoc = ({ schema, doc = {}, setErrors }) => {
  if (!schema) {
    return true;
  }

  const validationContext = schema.newContext();
  const validation = validationContext.validate(doc);

  const docErrors = validationContext.validationErrors();
  setErrors({ ...docErrors, submitAttempt: true });
  return validation;
};

export default validateDoc;
