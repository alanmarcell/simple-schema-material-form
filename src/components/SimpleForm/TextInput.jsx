import React from 'react';
// import PropTypes from 'prop-types';
import { omit, pathOr } from 'ramda';
// import { compose, withStateHandlers, pure } from 'recompact';
import { TextField } from 'material-ui';


const TextInput = props => {
  const { doc, setDoc, fieldName } = props;
  const value = pathOr('', [fieldName], doc);
  const cProps = omit(['fieldName', 'setDoc', 'doc'], props);

  const inputProps = {
    ...cProps,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };
  return <TextField {...inputProps} />;
};

export default TextInput;
