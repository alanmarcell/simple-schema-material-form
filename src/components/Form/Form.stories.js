import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withState } from '@dump247/storybook-state';
import { merge } from 'ramda';
import Typography from 'material-ui/Typography';
import SimpleSchema from 'simpl-schema';

import { SimpleForm } from './Form';
import TextInput from '../TextInput';
import Checkbox from '../Checkbox';
import Radio from '../Radio';
import withTests from '../../stories/withTests';

const setDoc = store => doc => store.set({ doc: merge(store.state.doc, doc) });

const Form = ({ store, children }) => (
  <SimpleForm
    {...store.state}
    setDoc={setDoc(store)}
    onSubmit={action('doc')}
  >
    {children}
  </SimpleForm>
);

const SessionSchema = new SimpleSchema({
  required: {
    type: String,
    min: 6,
  },
});

storiesOf('Form|TextInput', module)
  .addDecorator(withTests('Form'))
  .add('with TextInput Schema Validator', withState({ doc: {} }, (store) => (
    <div style={{ display: 'flex', alignItems: 'space-between', flexDirection: 'column' }}>
      <Typography variant="title" gutterBottom>
        If you hit submit without the required field a error is displayed
      </Typography>
      <Form store={store} schema={SessionSchema} >
        <TextInput fieldName="required" />
        <TextInput fieldName="optional" />
      </Form>
    </div>
  )))
  .add('with TextInput', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <TextInput fieldName="test" />
    </Form>
  )))
  .add('with 2 TextInput', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <TextInput fieldName="test" />
      <TextInput fieldName="test2" />
    </Form>
  )));


storiesOf('Form|Checkbox', module)
  .addDecorator(withTests('Form'))
  .add('with Checkbox', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <Checkbox selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
    </Form>
  )))
  .add('with Checkbox and Text Iput', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <Checkbox selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
      <TextInput fieldName="test2" />
    </Form>
  )));

storiesOf('Form|Radio', module)
  .addDecorator(withTests('Form'))
  .add('with Radio', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <Radio selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
    </Form>
  )))
  .add('with Checkbox and Text Iput', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <Radio selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
      <TextInput fieldName="test2" />
    </Form>
  )));
