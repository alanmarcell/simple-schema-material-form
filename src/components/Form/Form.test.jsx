import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import Form from './Form';
import TextInput from '../TextInput';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Form name="World" />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders without crashing', () => {
  const component = shallow(<Form name="World" />);
  component.dive();
  expect(component).toMatchSnapshot();
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Form name="World" ><TextInput fieldName="hello" /></Form>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders with a TextInput', () => {
  const component = shallow(<Form name="World" ><TextInput /></Form>);
  component.dive().dive();
  expect(component).toMatchSnapshot();
});

