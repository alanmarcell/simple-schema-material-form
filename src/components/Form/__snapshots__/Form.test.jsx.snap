// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with a TextInput 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withState(Form)
    name="World"
>
    <withState(TextInput) />
</withState(Form)>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <withState(TextInput) />,
      "doc": Object {},
      "name": "World",
      "setDoc": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <withState(TextInput) />,
        "doc": Object {},
        "name": "World",
        "setDoc": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withState(Form)
    name="World"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": null,
      "doc": Object {},
      "name": "World",
      "setDoc": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": null,
        "doc": Object {},
        "name": "World",
        "setDoc": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
